src/Nethermind/Nethermind.Xdc/EpochSwitchManager.cs

```diff
@@ -0,0 +1,454 @@
+            }
+        }
+
+        return TryBinarySearchBlockByEpochNumber(tempTCEpoch, estBlockNum, epochSwitchInfo.EpochSwitchBlockInfo.BlockNumber, (ulong)xdcSpec.SwitchBlock, (ulong)xdcSpec.EpochLength, out epochBlockInfo);
+    }
+
+    private bool TryGetBlockInfoInCache(ulong estRound, ulong epoch, out BlockInfo epochBlockInfo)
+    {
+        var epochSwitchInCache = new List<BlockInfo>();
+
+        for (ulong r = estRound; r < estRound + (ulong)epoch; r++)
+        {
+            if (_round2EpochBlockInfo.TryGetValue(r, out BlockInfo blockInfo) && blockInfo is not null)
+            {
+                epochSwitchInCache.Add(blockInfo);
+            }
+        }
+
+        if (epochSwitchInCache.Count == 1)
+        {
+            epochBlockInfo = epochSwitchInCache[0];
+            return true;
+        }
+        else if (epochSwitchInCache.Count == 0)
+        {
+            epochBlockInfo = null;
+            return false;
+        }
+
+        foreach (var blockInfo in epochSwitchInCache)
+        {
+            var header = _tree.FindHeader(blockInfo.BlockNumber);
+            if (header is null)
+            {
+                continue;
+            }
+            if (header.Hash == blockInfo.Hash)
+            {
+                epochBlockInfo = blockInfo;
+                return true;
+            }
+        }
+
+        epochBlockInfo = null;
+        return false;
+    }
+
+    private bool TryBinarySearchBlockByEpochNumber(ulong tempTCEpoch, long start, long end, ulong switchBlock, ulong epoch, out BlockInfo epochBlockInfo)
+    {
+        while (start < end)
+        {
+            var header = (XdcBlockHeader)_tree.FindHeader((start + end) / 2);
+            if (header is null)
+            {
+                epochBlockInfo = null;
+                return false;
+            }
+
+            bool isEpochSwitch = IsEpochSwitchAtBlock(header, out ulong epochNum);
+
+            if (epochNum == tempTCEpoch)
+            {
+                if (!Utils.TryGetExtraFields(header, (long)switchBlock, out var extraFieldsV2, out _))
+                {
+                    epochBlockInfo = null;
+                    return false;
+                }
+
+                ulong round = extraFieldsV2.CurrentRound;
+
+                if (isEpochSwitch)
+                {
+                    epochBlockInfo = new BlockInfo(header.Hash, round, header.Number);
+                    return true;
+                }
+                else
+                {
+                    end = header.Number;
+                    // trick to shorten the search
+                    ulong estStart = (ulong)end - (round % epoch);
+                    if (start < (long)estStart)
+                    {
+                        start = (long)estStart;
+                    }
+                }
+            }
+            else if (epochNum > tempTCEpoch)
+            {
+                end = header.Number;
+            }
+            else if (epochNum < tempTCEpoch)
```

Comment:
The condition can be omitted here, as this is the only possible case left, and just leave it as `else`